File: web/src/test/java/no/nav/sbl/dialogarena/rest/ressurser/personalia/NavEnhetRessursTest.java
package no.nav.sbl.dialogarena.rest.ressurser.personalia;

import no.nav.sbl.dialogarena.rest.ressurser.personalia.NavEnhetRessurs.NavEnhetFrontend;
import no.nav.sbl.dialogarena.sendsoknad.domain.adresse.AdresseForslag;
import no.nav.sbl.dialogarena.sendsoknad.domain.norg.NavEnhet;
import no.nav.sbl.dialogarena.sendsoknad.domain.oidc.StaticSubjectHandlerService;
import no.nav.sbl.dialogarena.sendsoknad.domain.oidc.SubjectHandler;
import no.nav.sbl.dialogarena.sendsoknad.domain.util.KommuneTilNavEnhetMapper;
import no.nav.sbl.dialogarena.sikkerhet.Tilgangskontroll;
import no.nav.sbl.dialogarena.soknadinnsending.business.service.SoknadsmottakerService;
import no.nav.sbl.dialogarena.sendsoknad.domain.digisosapi.KommuneInfoService;
import no.nav.sbl.dialogarena.soknadinnsending.consumer.norg.NorgService;
import no.nav.sbl.soknadsosialhjelp.soknad.adresse.JsonAdresse;
import no.nav.sbl.soknadsosialhjelp.soknad.adresse.JsonAdresseValg;
import no.nav.sbl.soknadsosialhjelp.soknad.adresse.JsonGateAdresse;
import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonKilde;
import no.nav.sbl.soknadsosialhjelp.soknad.JsonSoknadsmottaker;
import no.nav.sbl.soknadsosialhjelp.soknad.personalia.JsonPersonalia;
import no.nav.sbl.sosialhjelp.domain.SoknadUnderArbeid;
import no.nav.sbl.sosialhjelp.soknadunderbehandling.SoknadUnderArbeidRepository;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static no.nav.sbl.dialogarena.sendsoknad.domain.oidc.OidcFeatureToggleUtils.IS_RUNNING_WITH_OIDC;
import static no.nav.sbl.dialogarena.soknadinnsending.business.service.soknadservice.SoknadService.createEmptyJsonInternalSoknad;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.nullValue;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.*;

@RunWith(MockitoJUnitRunner.class)
public class NavEnhetRessursTest {

    private static final String BEHANDLINGSID = "123";

    private static final JsonAdresse OPPHOLDSADRESSE = new JsonGateAdresse()
            .withKilde(JsonKilde.BRUKER)
            .withType(JsonAdresse.Type.GATEADRESSE)
            .withLandkode("NOR")
            .withKommunenummer("123")
            .withAdresselinjer(null)
            .withBolignummer("1")
            .withPostnummer("2")
            .withPoststed("Oslo")
            .withGatenavn("Sanntidsgata")
            .withHusnummer("1337")
            .withHusbokstav("A");

    private static final String ENHETSNAVN = "NAV Testenhet";
    private static final String KOMMUNENAVN = "Test kommune";
    private static final String KOMMUNENR = KommuneTilNavEnhetMapper.getDigisoskommuner().get(0);
    private static final String ENHETSNR = "1234";
    private static final String ORGNR = "123456789";
    private static final String ENHETSNAVN_2 = "NAV Van";
    private static final String KOMMUNENAVN_2 = "Enummok kommune";
    private static final String KOMMUNENR_2 = KommuneTilNavEnhetMapper.getDigisoskommuner().get(1);
    private static final String ENHETSNR_2 = "5678";
    private static final String ORGNR_2 = "987654321";
    private static final JsonSoknadsmottaker SOKNADSMOTTAKER = new JsonSoknadsmottaker()
            .withNavEnhetsnavn(ENHETSNAVN + ", " + KOMMUNENAVN)
            .withEnhetsnummer(ENHETSNR)
            .withKommunenummer(KOMMUNENR);


    private static final JsonSoknadsmottaker SOKNADSMOTTAKER_2 = new JsonSoknadsmottaker()
            .withNavEnhetsnavn(ENHETSNAVN_2 + ", " + KOMMUNENAVN_2)
            .withEnhetsnummer(ENHETSNR_2)
            .withKommunenummer(KOMMUNENR_2);

    private static final AdresseForslag SOKNADSMOTTAKER_FORSLAG = new AdresseForslag();
    private static final AdresseForslag SOKNADSMOTTAKER_FORSLAG_2 = new AdresseForslag();

    private static final NavEnhet NAV_ENHET = new NavEnhet();
    private static final NavEnhet NAV_ENHET_2 = new NavEnhet();

    static {
        SOKNADSMOTTAKER_FORSLAG.geografiskTilknytning = ENHETSNAVN;
        SOKNADSMOTTAKER_FORSLAG.kommunenavn = KOMMUNENAVN;
        SOKNADSMOTTAKER_FORSLAG.kommunenummer = KOMMUNENR;

        NAV_ENHET.navn = ENHETSNAVN;
        NAV_ENHET.sosialOrgnr = ORGNR;
        NAV_ENHET.enhetNr = ENHETSNR;

        SOKNADSMOTTAKER_FORSLAG_2.geografiskTilknytning = ENHETSNAVN_2;
        SOKNADSMOTTAKER_FORSLAG_2.kommunenavn = KOMMUNENAVN_2;
        SOKNADSMOTTAKER_FORSLAG_2.kommunenummer = KOMMUNENR_2;

        NAV_ENHET_2.navn = ENHETSNAVN_2;
        NAV_ENHET_2.sosialOrgnr = ORGNR_2;
        NAV_ENHET_2.enhetNr = ENHETSNR_2;
    }

    private static final String EIER = "123456789101";

    @Mock
    private SoknadUnderArbeidRepository soknadUnderArbeidRepository;

    @Mock
    private Tilgangskontroll tilgangskontroll;

    @Mock
    private SoknadsmottakerService soknadsmottakerService;

    @Mock
    private NorgService norgService;

    @Mock
    private KommuneInfoService kommuneInfoService;

    @InjectMocks
    private NavEnhetRessurs navEnhetRessurs;

    @Before
    public void setUp() {
        SubjectHandler.setSubjectHandlerService(new StaticSubjectHandlerService());
        System.setProperty(IS_RUNNING_WITH_OIDC, "true");
    }

    @After
    public void tearDown() {
        SubjectHandler.resetOidcSubjectHandlerService();
        System.setProperty(IS_RUNNING_WITH_OIDC, "false");
    }

    @Test
    public void getNavEnheterSkalReturnereEnheterRiktigKonvertert(){
        SoknadUnderArbeid soknadUnderArbeid = new SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER));
        soknadUnderArbeid.getJsonInternalSoknad().getSoknad().withMottaker(SOKNADSMOTTAKER).getData().getPersonalia()
                .withOppholdsadresse(OPPHOLDSADRESSE.withAdresseValg(JsonAdresseValg.FOLKEREGISTRERT));
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(soknadUnderArbeid);
        when(soknadsmottakerService.finnAdresseFraSoknad(any(JsonPersonalia.class), eq("folkeregistrert"))).thenReturn(
                Arrays.asList(SOKNADSMOTTAKER_FORSLAG, SOKNADSMOTTAKER_FORSLAG_2));
        when(norgService.getEnhetForGt(ENHETSNAVN)).thenReturn(NAV_ENHET);
        when(norgService.getEnhetForGt(ENHETSNAVN_2)).thenReturn(NAV_ENHET_2);

        List<NavEnhetFrontend> navEnhetFrontends = navEnhetRessurs.hentNavEnheter(BEHANDLINGSID);

        assertThatEnheterAreCorrectlyConverted(navEnhetFrontends, Arrays.asList(SOKNADSMOTTAKER, SOKNADSMOTTAKER_2));
        assertThat(navEnhetFrontends.get(0).valgt, is(true));
        assertThat(navEnhetFrontends.get(1).valgt, is(false));
    }

    @Test
    public void getValgtNavEnhetSkalReturnereEnhetRiktigKonvertert(){
        SoknadUnderArbeid soknadUnderArbeid = new SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER));
        soknadUnderArbeid.getJsonInternalSoknad().getSoknad().withMottaker(SOKNADSMOTTAKER).getData().getPersonalia()
                .withOppholdsadresse(OPPHOLDSADRESSE.withAdresseValg(JsonAdresseValg.FOLKEREGISTRERT));
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(soknadUnderArbeid);

        NavEnhetFrontend navEnhetFrontend = navEnhetRessurs.hentValgtNavEnhet(BEHANDLINGSID);

        assertThatEnhetIsCorrectlyConverted(navEnhetFrontend, SOKNADSMOTTAKER);
        assertThat(navEnhetFrontend.valgt, is(true));
    }

    @Test
    public void getNavEnheterSkalReturnereTomListeNaarOppholdsadresseIkkeErValgt(){
        SoknadUnderArbeid soknadUnderArbeid = new SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER));
        soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getPersonalia()
                .withOppholdsadresse(OPPHOLDSADRESSE.withAdresseValg(null));
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(soknadUnderArbeid);
        when(soknadsmottakerService.finnAdresseFraSoknad(any(JsonPersonalia.class), eq(null))).thenReturn(new ArrayList<>());
        when(norgService.getEnhetForGt(ENHETSNAVN)).thenReturn(NAV_ENHET);
        when(norgService.getEnhetForGt(ENHETSNAVN_2)).thenReturn(NAV_ENHET_2);

        List<NavEnhetFrontend> navEnhetFrontends = navEnhetRessurs.hentNavEnheter(BEHANDLINGSID);

        assertTrue(navEnhetFrontends.isEmpty());
    }

    @Test
    public void getValgtNavEnhetSkalReturnereNullNarOppholdsadresseIkkeErValgt(){
        SoknadUnderArbeid soknadUnderArbeid = new SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER));
        soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getPersonalia()
                .withOppholdsadresse(OPPHOLDSADRESSE.withAdresseValg(null));
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(soknadUnderArbeid);

        NavEnhetFrontend navEnhetFrontends = navEnhetRessurs.hentValgtNavEnhet(BEHANDLINGSID);

        assertNull(navEnhetFrontends);
    }

    @Test
    public void putNavEnhetSkalSetteNavenhet(){
        SoknadUnderArbeid soknadUnderArbeid = new SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER));
        soknadUnderArbeid.getJsonInternalSoknad().getSoknad().withMottaker(SOKNADSMOTTAKER).getData().getPersonalia()
                .withOppholdsadresse(OPPHOLDSADRESSE.withAdresseValg(JsonAdresseValg.FOLKEREGISTRERT));
        when(soknadUnderArbeidRepository.hentSoknad(anyString(), anyString())).thenReturn(soknadUnderArbeid);
        doNothing().when(tilgangskontroll).verifiserAtBrukerKanEndreSoknad(anyString());
        NavEnhetFrontend navEnhetFrontend = new NavEnhetFrontend()
                .withEnhetsnavn(ENHETSNAVN_2)
                .withKommunenavn(KOMMUNENAVN_2)
                .withOrgnr(ORGNR_2)
                .withEnhetsnr(ENHETSNR_2);

        navEnhetRessurs.updateNavEnhet(BEHANDLINGSID, navEnhetFrontend);

        SoknadUnderArbeid updatedSoknadUnderArbeid = catchSoknadUnderArbeidSentToOppdaterSoknadsdata();
        JsonSoknadsmottaker jsonSoknadsmottaker = updatedSoknadUnderArbeid.getJsonInternalSoknad().getSoknad().getMottaker();
        assertThatEnhetIsCorrectlyConverted(navEnhetFrontend, jsonSoknadsmottaker);
    }

    private void assertThatEnheterAreCorrectlyConverted(List<NavEnhetFrontend> navEnhetFrontends, List<JsonSoknadsmottaker> jsonSoknadsmottakers){
        for (int i = 0; i < navEnhetFrontends.size(); i++){
            assertThatEnhetIsCorrectlyConverted(navEnhetFrontends.get(i), jsonSoknadsmottakers.get(i));
        }
    }

    private void assertThatEnhetIsCorrectlyConverted(NavEnhetFrontend navEnhetFrontend, JsonSoknadsmottaker soknadsmottaker) {
        if (navEnhetFrontend == null){
            assertThat(soknadsmottaker, nullValue());
            return;
        }

        String kombinertnavn = soknadsmottaker.getNavEnhetsnavn();
        String enhetsnavn = kombinertnavn.substring(0, kombinertnavn.indexOf(','));
        String kommunenavn = kombinertnavn.substring(kombinertnavn.indexOf(',') + 2);

        assertThat("Enhetsnavn", navEnhetFrontend.enhetsnavn, is(enhetsnavn));
        assertThat("kommunenavn", navEnhetFrontend.kommunenavn, is(kommunenavn));
        assertThat("enhetsnr", navEnhetFrontend.enhetsnr, is(soknadsmottaker.getEnhetsnummer()));
    }

    private SoknadUnderArbeid catchSoknadUnderArbeidSentToOppdaterSoknadsdata() {
        ArgumentCaptor<SoknadUnderArbeid> argument = ArgumentCaptor.forClass(SoknadUnderArbeid.class);
        verify(soknadUnderArbeidRepository).oppdaterSoknadsdata(argument.capture(), anyString());
        return argument.getValue();
    }
}

File: src/test/kotlin/no/nav/sosialhjelp/soknad/navenhet/NavEnhetRessursTest.kt
package no.nav.sosialhjelp.soknad.navenhet

import io.mockk.Called
import io.mockk.called
import io.mockk.clearAllMocks
import io.mockk.every
import io.mockk.just
import io.mockk.mockk
import io.mockk.mockkObject
import io.mockk.runs
import io.mockk.slot
import io.mockk.unmockkObject
import io.mockk.verify
import no.nav.sbl.soknadsosialhjelp.soknad.JsonSoknadsmottaker
import no.nav.sbl.soknadsosialhjelp.soknad.adresse.JsonAdresse
import no.nav.sbl.soknadsosialhjelp.soknad.adresse.JsonAdresseValg
import no.nav.sbl.soknadsosialhjelp.soknad.adresse.JsonGateAdresse
import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonKilde
import no.nav.sosialhjelp.soknad.app.MiljoUtils
import no.nav.sosialhjelp.soknad.app.exceptions.AuthorizationException
import no.nav.sosialhjelp.soknad.app.mapper.KommuneTilNavEnhetMapper
import no.nav.sosialhjelp.soknad.app.subjecthandler.StaticSubjectHandlerImpl
import no.nav.sosialhjelp.soknad.app.subjecthandler.SubjectHandlerUtils
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.SoknadUnderArbeid
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.SoknadUnderArbeidRepository
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.SoknadUnderArbeidStatus
import no.nav.sosialhjelp.soknad.innsending.SoknadService.Companion.createEmptyJsonInternalSoknad
import no.nav.sosialhjelp.soknad.navenhet.dto.NavEnhetFrontend
import no.nav.sosialhjelp.soknad.tilgangskontroll.Tilgangskontroll
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatExceptionOfType
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.time.LocalDateTime

internal class NavEnhetRessursTest {

    companion object {
        private const val BEHANDLINGSID = "123"
        private const val OPPHOLDSADRESSE_KOMMUNENR = "0123"
        private val OPPHOLDSADRESSE: JsonAdresse = JsonGateAdresse()
            .withKilde(JsonKilde.BRUKER)
            .withType(JsonAdresse.Type.GATEADRESSE)
            .withLandkode("NOR")
            .withKommunenummer(OPPHOLDSADRESSE_KOMMUNENR)
            .withAdresselinjer(null)
            .withBolignummer("1")
            .withPostnummer("2")
            .withPoststed("Oslo")
            .withGatenavn("Sanntidsgata")
            .withHusnummer("1337")
            .withHusbokstav("A")

        private const val ENHETSNAVN = "NAV Testenhet"
        private const val KOMMUNENAVN = "Test kommune"
        private val KOMMUNENR = KommuneTilNavEnhetMapper.digisoskommuner[0]
        private const val ENHETSNR = "1234"
        private const val ORGNR = "123456789"

        private val SOKNADSMOTTAKER = JsonSoknadsmottaker()
            .withNavEnhetsnavn("$ENHETSNAVN, $KOMMUNENAVN")
            .withEnhetsnummer(ENHETSNR)
            .withKommunenummer(KOMMUNENR)
        private const val EIER = "123456789101"
    }

    private val soknadUnderArbeidRepository: SoknadUnderArbeidRepository = mockk()
    private val tilgangskontroll: Tilgangskontroll = mockk()
    private val navEnhetService: NavEnhetService = mockk()

    private val navEnhetRessurs = NavEnhetRessurs(
        tilgangskontroll = tilgangskontroll,
        soknadUnderArbeidRepository = soknadUnderArbeidRepository,
        navEnhetService = navEnhetService
    )

    private val navEnhetFrontend = NavEnhetFrontend(
        orgnr = ORGNR,
        enhetsnr = ENHETSNR,
        enhetsnavn = ENHETSNAVN,
        kommunenavn = KOMMUNENAVN,
        kommuneNr = KOMMUNENR,
        behandlingsansvarlig = null,
        valgt = true,
        isMottakMidlertidigDeaktivert = false,
        isMottakDeaktivert = false
    )

    @BeforeEach
    internal fun setUp() {
        clearAllMocks()

        mockkObject(MiljoUtils)
        every { MiljoUtils.isNonProduction() } returns true
        SubjectHandlerUtils.setNewSubjectHandlerImpl(StaticSubjectHandlerImpl())
        every { tilgangskontroll.verifiserBrukerHarTilgangTilSoknad(any()) } just runs
    }

    @AfterEach
    internal fun tearDown() {
        SubjectHandlerUtils.resetSubjectHandlerImpl()
        unmockkObject(MiljoUtils)
    }

    @Test
    internal fun `getNavEnheter - skal returnere NavEnhetFrontend-liste`() {
        val soknadUnderArbeid = createSoknadUnderArbeid()
        soknadUnderArbeid.jsonInternalSoknad!!.soknad.withMottaker(SOKNADSMOTTAKER).data.personalia
            .withOppholdsadresse(OPPHOLDSADRESSE.withAdresseValg(JsonAdresseValg.SOKNAD))

        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns soknadUnderArbeid
        every { navEnhetService.getNavEnhet(any(), any(), any()) } returns navEnhetFrontend

        val response = navEnhetRessurs.getNavEnheter(BEHANDLINGSID)

        assertThat(response).hasSize(1)
        assertThat(response[0].valgt).isTrue
    }

    @Test
    fun `getNavEnheter - skal returnere tom liste`() {
        val soknadUnderArbeid = createSoknadUnderArbeid()
        soknadUnderArbeid.jsonInternalSoknad!!.soknad.withMottaker(SOKNADSMOTTAKER).data.personalia
            .withOppholdsadresse(OPPHOLDSADRESSE.withAdresseValg(JsonAdresseValg.SOKNAD))

        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns soknadUnderArbeid
        every { navEnhetService.getNavEnhet(any(), any(), any()) } returns null

        val response = navEnhetRessurs.getNavEnheter(BEHANDLINGSID)

        assertThat(response).isEmpty()
    }

    @Test
    fun `getNavEnheter - skal kaste feil`() {
        val soknadUnderArbeid: SoknadUnderArbeid = mockk()
        every { soknadUnderArbeid.jsonInternalSoknad?.soknad } returns null
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns soknadUnderArbeid

        assertThatExceptionOfType(IllegalStateException::class.java)
            .isThrownBy { navEnhetRessurs.getNavEnheter(BEHANDLINGSID) }

        verify { navEnhetService wasNot called }
    }

    @Test
    fun `getValgtNavEnhet - skal hente NavEnhetFrontend`() {
        val soknadUnderArbeid = createSoknadUnderArbeid()
        soknadUnderArbeid.jsonInternalSoknad!!.soknad.withMottaker(SOKNADSMOTTAKER).data.personalia
            .withOppholdsadresse(OPPHOLDSADRESSE.withAdresseValg(JsonAdresseValg.SOKNAD))

        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns soknadUnderArbeid
        every { navEnhetService.getValgtNavEnhet(SOKNADSMOTTAKER) } returns navEnhetFrontend

        val response = navEnhetRessurs.getValgtNavEnhet(BEHANDLINGSID)

        assertThat(response).isNotNull
        assertThat(response?.valgt).isTrue
    }

    @Test
    fun `getValgtNavEnhet - null hvis kommunenummer er null`() {
        val soknadUnderArbeid = createSoknadUnderArbeid()
        soknadUnderArbeid.jsonInternalSoknad!!.soknad
            .withMottaker(
                JsonSoknadsmottaker()
                    .withNavEnhetsnavn(ENHETSNAVN)
                    .withEnhetsnummer(ENHETSNR)
                    .withKommunenummer(null)
            )
            .data.personalia.withOppholdsadresse(OPPHOLDSADRESSE.withAdresseValg(JsonAdresseValg.SOKNAD))

        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns soknadUnderArbeid

        assertThat(navEnhetRessurs.getValgtNavEnhet(BEHANDLINGSID)).isNull()
    }

    @Test
    fun `getValgtNavEnhet - null hvis navEnhetsnavn er null`() {
        val soknadUnderArbeid = createSoknadUnderArbeid()
        soknadUnderArbeid.jsonInternalSoknad!!.soknad
            .withMottaker(
                JsonSoknadsmottaker()
                    .withNavEnhetsnavn(null)
                    .withEnhetsnummer(ENHETSNR)
                    .withKommunenummer(KOMMUNENR)
            )
            .data.personalia.withOppholdsadresse(OPPHOLDSADRESSE.withAdresseValg(JsonAdresseValg.SOKNAD))

        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns soknadUnderArbeid

        assertThat(navEnhetRessurs.getValgtNavEnhet(BEHANDLINGSID)).isNull()
    }

    @Test
    fun `getValgtNavEnhet - skal kaste feil`() {
        val soknadUnderArbeid: SoknadUnderArbeid = mockk()
        every { soknadUnderArbeid.jsonInternalSoknad?.soknad?.mottaker } returns null
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns soknadUnderArbeid

        assertThatExceptionOfType(IllegalStateException::class.java)
            .isThrownBy { navEnhetRessurs.getValgtNavEnhet(BEHANDLINGSID) }

        verify { navEnhetService wasNot called }
    }

    @Test
    fun `putNavEnhet - skal oppdatere JsonSoknadsmottaker`() {
        val soknadUnderArbeid = createSoknadUnderArbeid()

        every { tilgangskontroll.verifiserAtBrukerKanEndreSoknad(any()) } just runs
        every { soknadUnderArbeidRepository.hentSoknad(any<String>(), any()) } returns soknadUnderArbeid

        val slot = slot<SoknadUnderArbeid>()
        every { soknadUnderArbeidRepository.oppdaterSoknadsdata(capture(slot), any()) } just runs

        navEnhetRessurs.putNavEnhet(BEHANDLINGSID, navEnhetFrontend)

        val oppdatertSoknadUnderArbeid = slot.captured

        val mottaker = oppdatertSoknadUnderArbeid.jsonInternalSoknad?.mottaker
        assertThat(mottaker).isNotNull
        assertThat(mottaker?.navEnhetsnavn).contains(navEnhetFrontend.enhetsnavn, navEnhetFrontend.kommunenavn)
        assertThat(mottaker?.organisasjonsnummer).isEqualTo(navEnhetFrontend.orgnr)

        val soknadsmottaker = oppdatertSoknadUnderArbeid.jsonInternalSoknad!!.soknad.mottaker
        val kombinertnavn = soknadsmottaker.navEnhetsnavn
        val enhetsnavn = kombinertnavn.substring(0, kombinertnavn.indexOf(','))
        val kommunenavn = kombinertnavn.substring(kombinertnavn.indexOf(',') + 2)
        assertThat(navEnhetFrontend.enhetsnavn).isEqualTo(enhetsnavn)
        assertThat(navEnhetFrontend.kommunenavn).isEqualTo(kommunenavn)
        assertThat(navEnhetFrontend.enhetsnr).isEqualTo(soknadsmottaker.enhetsnummer)
    }

    @Test
    internal fun `hentNavEnheter skal kaste AuthorizationException ved manglende tilgang`() {
        every { tilgangskontroll.verifiserBrukerHarTilgangTilSoknad(any()) } throws AuthorizationException("Not for you my friend")

        assertThatExceptionOfType(AuthorizationException::class.java)
            .isThrownBy { navEnhetRessurs.getNavEnheter(BEHANDLINGSID) }

        verify { soknadUnderArbeidRepository wasNot Called }
    }

    @Test
    internal fun `hentValgtNavEnhet skal kaste AuthorizationException ved manglende tilgang`() {
        every { tilgangskontroll.verifiserBrukerHarTilgangTilSoknad(any()) } throws AuthorizationException("Not for you my friend")

        assertThatExceptionOfType(AuthorizationException::class.java)
            .isThrownBy { navEnhetRessurs.getValgtNavEnhet(BEHANDLINGSID) }

        verify { soknadUnderArbeidRepository wasNot Called }
    }

    @Test
    internal fun `updateNavEnhet skal kaste AuthorizationException ved manglende tilgang`() {
        every { tilgangskontroll.verifiserAtBrukerKanEndreSoknad(any()) } throws AuthorizationException("Not for you my friend")

        val navEnhetFrontend = mockk<NavEnhetFrontend>()

        assertThatExceptionOfType(AuthorizationException::class.java)
            .isThrownBy { navEnhetRessurs.putNavEnhet(BEHANDLINGSID, navEnhetFrontend) }

        verify { soknadUnderArbeidRepository wasNot Called }
    }

    private fun createSoknadUnderArbeid(): SoknadUnderArbeid {
        return SoknadUnderArbeid(
            versjon = 1L,
            behandlingsId = BEHANDLINGSID,
            tilknyttetBehandlingsId = null,
            eier = EIER,
            jsonInternalSoknad = createEmptyJsonInternalSoknad(EIER),
            status = SoknadUnderArbeidStatus.UNDER_ARBEID,
            opprettetDato = LocalDateTime.now(),
            sistEndretDato = LocalDateTime.now()
        )
    }
}
